inventory/byo/hosts.origin.example

```diff
@@ -89,6 +89,21 @@ openshift_release=v1.4
 # Skip upgrading Docker during an OpenShift upgrade, leaves the current Docker version alone.
 # docker_upgrade=False
 
+
+# Upgrade Hooks
+#
+# Hooks are available to run custom tasks at various points during a cluster
+# upgrade. Each hook should point to a file with Ansible tasks defined. Suggest using
+# absolute paths, if not the path will be treated as relative to the file where the
+# hook is actually used.
+#
+# Tasks to run before each master is upgraded:
+# openshift_upgrade_pre_master_hook=/usr/share/custom/pre_master.yml
```

Comment:
> The hook runs against each master in serial. It can contain only tasks as it needs to plug into the middle of our serial upgrade.

I think that definitely needs to be documented. I can see some issues with this approach, since the master might not be able to reach the lb management interface or may not have access to the credentials needed. Of course that can be solved by using delegated tasks and possibly by setting variables prior to running the upgrade playbook, but I think it is important to document the behavior.

> I don't think we can set an interface for what variables are set at this point, the caller is going to have to do their own thing. If they choose to use something of ours, there are presently no guarantees it will continue working going forward.

I see this as a potential problem, but as long as we document this limitation, then at least the user has been warned.

> Hook failure = upgrade failure, you will have to solve and re-run. Hooks will need to be able to accommodate that as upgrade does.

Keep in mind that we'll need to provide some examples, so that users understand how to achieve that behavior. It would be bad if their hook consistently fails to remove a host from the LB that is already removed for example.

More importantly, how would it be handled if the 3rd host out of 3 was where the failure occurred after the 3rd host was removed from the LB? Would the 1st and 2nd host be removed from the LB prior to getting back to the 3rd host (which is already removed from the LB)?

I'm thinking we are going to have to store some state on the hosts (using ansible local facts) so that we can track the state of the upgrade. If we go that route, we may need to allow the hooks to have some say in setting their state, to allow for users to tweak hooks part way through a run and have that affect the system state.

> I do not expect hooks to return data, but they can set variables on hosts usable by your later hooks, that seems fine to me.

At some point we may have a need for hooks to inject data back into the upgrade. Simple LB operations may not require it, but something more complex like a SDN hook might. At the very least we should document that hooks are not able to affect the upgrade state today.

> Does this sound ok, and if so where should I document it? At this point I think we have to keep it simple and expand on things later as need arises. I kind of need to see how it plays out for Ops needs as well. (working on that now)

I think documenting it in the upgrade README would be a good place until we can get it ironed out and added to the upstream docs.